src/monitoring_tests/cost_coverage_per_solver_test.py

```diff
@@ -0,0 +1,104 @@
+        self.total_coverage_per_solver: Dict[str, float] = {}
+        self.original_block = self.web3_api.get_current_block_number()
+
+    def cost_coverage(self, competition_data: dict[str, Any], gas_cost: float) -> bool:
+        """
+        This function compares the fees, as perceived by the winning solver, and the actual
+        execution cost of the corresponding settlement. This is refered to as cost_coverage,
+        and is supposed to monitor how well the fees end up approximating the execution cost
+        of a solution.
+        We also look at the coverage, from the protocol perspective, of each settlement, i.e.,
+        we compare the payout made to the solver with the fees collected.
+        """
+
+        solution = competition_data["solutions"][-1]
+        fees = float(int(solution["objective"]["fees"]) / 10**18)
+        surplus = float(int(solution["objective"]["surplus"]) / 10**18)
+        solver = solution["solver"]
+        ref_score = 0
+        if len(competition_data["solutions"]) > 1:
+            second_best_sol = competition_data["solutions"][-1]
+            if "score" in second_best_sol:
+                ref_score = second_best_sol["score"]
+            elif "scoreDiscounted" in second_best_sol:
+                ref_score = second_best_sol["scoreDiscounted"]
+            else:
+                ref_score = second_best_sol["scoreProtocol"]
+        if solver in self.cost_coverage_per_solver:
+            self.cost_coverage_per_solver[solver] += fees - gas_cost
+            self.total_coverage_per_solver[solver] += ref_score - surplus
+        else:
+            self.cost_coverage_per_solver[solver] = fees - gas_cost
+            self.total_coverage_per_solver[solver] = ref_score - surplus
+
+        return True
+
+    def run(self, tx_hash: str) -> bool:
+        """
+        Wrapper function for the whole test. Checks if solver competition data is retrievable
+        and runs test, else returns False to add to list of unchecked hashes.
+        """
+        solver_competition_data = self.orderbook_api.get_solver_competition_data(
+            tx_hash
+        )
+        transaction = self.web3_api.get_transaction(tx_hash)
+        receipt = self.web3_api.get_receipt(tx_hash)
+        gas_cost = 0.0
+        if transaction is not None and receipt is not None:
+            gas_used, gas_price = self.web3_api.get_batch_gas_costs(
+                transaction, receipt
+            )
+            gas_cost = float(gas_used) * float(gas_price) / 10**18
+        if gas_cost == 0 or solver_competition_data is None:
+            return False
+
+        success = self.cost_coverage(solver_competition_data, gas_cost)
+
+        ### This part takes care of the reporting once a day.
+        current_block = self.web3_api.get_current_block_number()
+        if current_block is None or self.original_block is None:
+            return success
+        if current_block - self.original_block > DAY_BLOCK_INTERVAL:
```

Comment:
We could also loop through all solvers here to get a daily overview over everything.
Otherwise we might get that report with widely different time horizons for less successful solvers.